### ISSUE TITLE
Add request scoped context manager based resources for the fastapi integration

### ISSUE THREAD

**Issue by meadsteve** [meadsteve/lagom#180]
Similar to: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/

See conversation on this issue: https://github.com/meadsteve/lagom/issues/179#issuecomment-979763975

Stretch goal: make this generic enough to use for all lagom usages.

**Comment by meadsteve** [comment]
@rowanbarrie this was slightly easier to hook into than I thought.
I'm currently experimenting with an interface like this:

```python
# Define a contextmanager for the dependency. The __exit__ (finally in this case) will happen after the request has loaded.
@dependency_definition(container)
@contextmanager
def _load_then_clean() -> Iterator[SomeThing]:
    try:
        yield SomeThing()
    finally:
        SomeThing.do_some_cleanup()

# This type then needs to be registered as a managed singleton:
deps = FastApiIntegration(
    container, request_context_singletons=[SomeThing]
)
```

Based around defining a ContextManager for the thing required and then the framework integration handles calling __exit__ at the end of the request handling.

**Comment by meadsteve** [comment]
released in lagom 1.7.0
